finality-aleph/src/data_io/data_interpreter.rs

```diff
@@ -0,0 +1,132 @@
+                // 1) It came from our node and thus was prepared based on our local `best_block`.
+                // 2) It came from outside of our node and went from DataStore, thus the only way it was let through is
+                //    by reaching the `Available` status. Thus it must be available.
+
+                // There are malicious examples of hopeless forks (for which the branch is incorrect) which if input
+                // to the below procedure would (in some cases) yield a result of `Some(b)` instead of `None`. That's why
+                // we must make sure here that not only the data is available, but also that it is not a hopeless fork.
+
+                // Note also that there might be cases for which prior to entering AlephBFT a proposal could have got status
+                // Available(BranchOk) and now the same proposal would be Available(HopelessFork). This is still fine as
+                // the code below will handle such proposals correctly (due to the fact that the branch in such a proposal
+                // is perfectly honest and thus it will be detected as conflicting with previously finalized blocks).
+
+                // WARNING: If we ever enable pruning, this code (and the code in Data Store) must be carefully analyzed
+                // for possible safety violations.
+
+                use crate::data_io::proposal::{AvailableKind::*, ProposalStatus::*};
+                let status = self
+                    .proposal_status_provider
+                    .lock()
+                    .get_status(&proposal, None);
+                match status {
+                    Available(BranchOk) => {
+                        if proposal.number_top_block() <= self.last_finalized_by_aleph.num {
+                            None
+                        } else if let Some(block) =
+                            proposal.block_at_num(self.last_finalized_by_aleph.num)
+                        {
+                            // We must check if the last_finalized_by_aleph block agrees with the one in the proposed branch
+                            // Otherwise this is a hopeless fork that we must ignore.
+                            if block.hash == self.last_finalized_by_aleph.hash {
+                                Some(proposal.top_block())
+                            } else {
+                                None
+                            }
+                        } else {
+                            // The last_finalized_by_aleph block is strictly below the branch bottom block. We apply the proposal,
+                            // as it agrees with last_finalized_by_aleph.
+                            Some(proposal.top_block())
+                        }
+                    }
+                    Available(HopelessFork) => {
+                        debug!(target:"afa", "HopelessFork {:?} encountered in Data. Skipping.", proposal);
```

Comment:
what about introducing another target for this part of code/module?

Reply:
In general, yes, but I beg you, not in this PR, as I'm desperately wanting to get it over with ASAP :)